go.mod

```diff
@@ -274,3 +274,5 @@ require (
 	gopkg.in/yaml.v3 v3.0.1 // indirect
 	lukechampine.com/blake3 v1.3.0 // indirect
 )
+
+replace github.com/libp2p/go-libp2p => github.com/libp2p/go-libp2p v0.37.1-0.20241202220543-9024f8e8c86e
```

Comment:
There seem to be a regression between  9b887ea (this PR) and  433444b (latest master) caused by TCP reuse logic in go-libp2p bumped here.

Repro:

- Start `ipfs daemon` 
- Open http://127.0.0.1:5001/webui#/peers and wait a minute:
  - this will trigger connections to peers to fetch any missing block get for geoip data (if not cached locally) – acting like accelerator for any underlying bugs in TCP port reuse
  - 💢 `staging-2024-12-04-9b887e` (go-libp2p master `v0.37.1-0.20241202220543-9024f8e8c86e`)  will grow CPU use once ipfs-webui Peers screen is opened. CPU peaks within 1minute.
  - 💚 `master-2024-12-03-433444b` (`go-libp2p v0.37.2`) does not trigger high CPU load

Here are docker images for repro convenience:
  - `docker run --rm -it --net=host ipfs/kubo:staging-2024-12-04-9b887ea`
  - `docker run --rm -it --net=host ipfs/kubo:master-2024-12-03-433444b`

I did `curl localhost:5001/debug/pprof/profile > ipfs.cpuprof ` and CPU profile seems to confirm the source is TCP reuse from https://github.com/libp2p/go-libp2p/pull/2984? (cc @aschmahmann  @MarcoPolo):

> ![cpu](https://github.com/user-attachments/assets/345ea491-507b-47b2-8d6a-19f3821b6ede)

Reply:
Related work:
- https://github.com/libp2p/go-libp2p/pull/3080
- https://github.com/libp2p/go-libp2p/pull/3081

Reply:
I am debugging sharness with `make -O -j 1`  because CI runs with `-j 10` and logs are mangled.

Reply:
Getting bizzare. Tests with the same go-libp2p and boxo version pass in https://github.com/ipfs/kubo/pull/10631.


The only `go.mod` diff between https://github.com/ipfs/kubo/pull/10565/commits/a436f4eda2055b69a7879d6af2da4c4028b48e2c (this PR, [ci failing due to timeout after 20m](https://github.com/ipfs/kubo/actions/runs/12420379556/job/34677757355)) and  https://github.com/ipfs/kubo/pull/10631/commits/070e6aed2a6b469e3062239e8029993efc4d851b (#10631, [ci green, sharness run takes 5minutes](https://github.com/ipfs/kubo/actions/runs/12420385789/job/34677776535)), is:

```diff
-       github.com/ipshipyard/p2p-forge v0.1.0
+       github.com/ipshipyard/p2p-forge v0.2.0
```

Reply:
Sharness passes green if we disable `libp2p.ShareTCPListener()` (https://github.com/ipfs/kubo/pull/10565/commits/b5cfd6d3b46aa8fe735a193fed4146a975ef0519 passed).

Something related to `libp2p.ShareTCPListener()` makes sharness hang and timeout.

Reply:
Run two tests to narrow down the search:
- :green_circle:  https://github.com/ipfs/kubo/pull/10565/commits/7a4ec80fcdd86aa2afd99a2c1029a6343aa9b163 disabled `libp2p.ShareTCPListener()` and sharness finished in 8m successfully ([log](https://github.com/ipfs/kubo/actions/runs/12433050915))
- :red_circle:  https://github.com/ipfs/kubo/pull/10565/commits/858e10accc86d7b95efac395bed2ba2cf9ea4c9c enabled `libp2p.ShareTCPListener()` and sharness  timeouted after 20m ([log](https://github.com/ipfs/kubo/actions/runs/12433463669))

Maybe increasing timeout would help, but taking 2x as long feels like a bug.

Reply:
Ok, running sequentially (https://github.com/ipfs/kubo/pull/10565#discussion_r1894060466)  found the problem:
- Enabling `libp2p.ShareTCPListener()` break PNET and `./t0181-private-network.sh` hangs – kinda makes sense, PNET is TCP-only feature of libp2p and we mess up with TCP. 

I'll now confirm this is the only one failing, 
and if other tests pass, will disable TCP sharing when PNET is enabled for now.

Reply:
Good news, PNET was the only reason sharness failed. Disabled port sharing when PNET is enabled (https://github.com/ipfs/kubo/pull/10565/commits/bb87df3ba3cec852560face6627003f349e7712a) and CI is green again:

- https://github.com/ipfs/kubo/actions/runs/12435566625 passed.

I'm going forward with Kubo 0.33.0-rc1.